pages/instructorGradebook/instructorGradebook.js

```diff
@@ -71,7 +71,15 @@ router.post('/', function(req, res, next) {
         ];
         sqldb.call('assessment_instances_update_score_perc', params, function(err, _result) {
             if (ERR(err, next)) return;
-            res.redirect(req.originalUrl);
+
+            params = {
+                assessment_instance_id: req.body.assessment_instance_id,
```

Comment:
Data in `req.body` is untrusted, so it's possible that this `assessment_instance_id` is one that the instructor does not have permission to modify. We should actually be checking this above in `assessment_instances_update_score_perc()`. Our standard approach to this would be to pass `res.locals.assessment.id` (trusted, because it's in `res.locals`) to `assessment_instances_update_score_perc()` and inside `assessment_instances_update_score_perc()` we first check whether the provided `assessment_instance_id` belongs to the `assessment_id` (throwing a access exception if not). After that we can trust that the `assessment_instance_id` is valid.

Reply:
The grade update was not modified, only the retrieval of the updated grade, so the vulnerability was there already. That said, I'm happy to update the sproc, though this would also require updating it on the other places where this sproc is called: the assessment instance log, and the upload by assessment. These other two places do not have the vulnerability (the log uses the instance from `res.locals`, score upload searches for the assessment instance ID from the assessment ID).

Reply:
Sorry for not being clearer --- this was absolutely a pre-existing vulnerability that I only noticed now because of your nearby changes 😄

If the other call sites have easy access to the `assessment_id` then I'd suggest just passing it in and doing the check in all cases. If that's a pain, then an alternative is to make the `assessment_id` argument optional for `assessment_instances_update_score_perc()` and only do the check if we get the argument passed in.

Reply:
assessment_id is not in `res.locals`, since this is not an assessment-related page... Should I use the course instance instead?

Reply:
Yep, good call. Another option here is to do something like this: https://github.com/PrairieLearn/PrairieLearn/blob/f37c625f9c67f2037d57277326d2e68927b9d9d4/pages/instructorAssessmentInstances/instructorAssessmentInstances.js#L83

This is somewhat less efficient because it adds another DB round trip, but we already do multiple queries for auth so one more isn't too bad.

Reply:
Yeah, that would probably be simpler. Should I do this "checkBelongs" version for course instance vs assessment instance in the gradebook page itself or should I create a new library for it? Or are you aware of a library function that already exists for this test?

Reply:
I don't think we have a `checkBelongs()` for an `assessment_instance_id` in a `course_instance_id`. Making a library function for this would be great. I don't have a great suggestion for an existing file to put it in, so maybe a new `lib/course.js`? (mimicking `lib/assessment.js` and `lib/question.js`)

Reply:
Would `lib/courseInstance.js` make more sense?

Reply:
Yes, except that I was thinking about the analogy with `lib/assessment.js` and `lib/question.js`. We don't have `lib/assessmentInstance.js`, for example. But I don't have a strong opinion here, so go with whatever you think is most natural.